cpp/apidoc/tutorials/row_wise_conversion.md

```diff
@@ -20,175 +20,175 @@
-// the underlying memory regions in-place. At the moment, arrow::jemalloc is only
-// supported on Unix systems, not Windows.
+#include <arrow/api.h>
 
 using arrow::DoubleBuilder;
 using arrow::Int64Builder;
 using arrow::ListBuilder;
 
-arrow::MemoryPool* pool = arrow::default_memory_pool();
-Int64Builder id_builder(pool);
-DoubleBuilder cost_builder(pool);
-std::unique_ptr<DoubleBuilder> components_values_builder(new DoubleBuilder(pool));
-ListBuilder components_builder(pool, std::move(components_values_builder));
-```
-
-Now we can loop over our existing data and insert it into the builders. The
-`Append` calls here may fail (e.g. we cannot allocate enough additional memory).
-Thus we need to check their return values. For more information on these values,
-check the documentation about `arrow::Status`.
+// While we want to use columnar data structures to build efficient operations, we
+// often receive data in a row-wise fashion from other systems. In the following,
+// we want give a brief introduction into the classes provided by Apache Arrow by
+// showing how to transform row-wise data into a columnar table.
+//
+// The data in this example is stored in the following struct:
+struct data_row {
+  int64_t id;
+  double cost;
+  std::vector<double> cost_components;
+};
 
-```
-for (const data_row& row : rows) {
-    ARROW_RETURN_NOT_OK(id_builder.Append(row.id));
-    ARROW_RETURN_NOT_OK(cost_builder.Append(row.cost));
+// Transforming a vector of structs into a columnar Table.
+//
+// The final representation should be an `arrow::Table` which in turn is made up of
+// an `arrow::Schema` and a list of `arrow::Column`. An `arrow::Column` is again a
+// named collection of one or more `arrow::Array` instances. As the first step, we
+// will iterate over the data and build up the arrays incrementally. For this task,
+// we provide `arrow::ArrayBuilder` classes that help in the construction of the
+// final `arrow::Array` instances.
+//
+// For each type, Arrow has a specially typed builder class. For the primitive
+// values `id` and `cost` we can use the respective `arrow::Int64Builder` and
+// `arrow::DoubleBuilder`. For the `cost_components` vector, we need to have two
+// builders, a top-level `arrow::ListBuilder` that builds the array of offsets and
+// a nested `arrow::DoubleBuilder` that constructs the underlying values array that
+// is referenced by the offsets in the former array.
+arrow::Status VectorToColumnarTable(const std::vector<struct data_row>& rows,
+                                    std::shared_ptr<arrow::Table>* table) {
+  // The builders are more efficient using
+  // arrow::jemalloc::MemoryPool::default_pool() as this can increase the size of
+  // the underlying memory regions in-place. At the moment, arrow::jemalloc is only
+  // supported on Unix systems, not Windows.
+  arrow::MemoryPool* pool = arrow::default_memory_pool();
+
+  auto id_builder = std::make_shared<Int64Builder>(pool);
```

Comment:
These builders are only used in the local scope. We should make them simple stack variables without wrapping them into `std::shared_ptr`.

Reply:
I wrapped them out of uniformity. The `cost_components_builder` has to be in a shared_ptr to be passed to the `components_builder`. The previous version was wrongly using a std::move then referencing again the (moved out) builder.

Our choice is between:
- 3 local and 1 shared_ptr builder
- 4 shared_ptr builders

I can revert this choice.